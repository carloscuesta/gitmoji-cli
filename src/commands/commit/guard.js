import chalk from 'chalk'

const errors = {
  scope: chalk.red('Enter a valid scope'),
  title: chalk.red('Enter a valid commit title'),
  message: chalk.red('Enter a valid commit message')
}

const title = (title: string) =>
  !title || title.includes('`') ? errors.title : true

const message = (message: string) =>
  message.includes('`') ? errors.message : true

const scope = (scope: string) => (scope.includes('`') ? errors.scope : true)

export default {
  message,
  scope,
  title
}
