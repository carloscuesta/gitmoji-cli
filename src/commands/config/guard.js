import chalk from 'chalk'
import isURL from 'validator/lib/isURL.js'

const errors = {
  url: chalk.red('Enter a valid API URL')
}

const url = (url: string) => (isURL(url) ? true : errors.url)

export default {
  url
}
