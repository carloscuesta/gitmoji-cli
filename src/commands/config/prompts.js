// @flow
import configurationVault from '../../utils/configurationVault'

export const CONFIGURATION_PROMPT_NAMES = {
  AUTO_ADD: 'autoAdd',
  EMOJI_FORMAT: 'emojiFormat',
  SCOPE_PROMPT: 'scopePrompt',
  SIGNED_COMMIT: 'signedCommit'
}

export const EMOJI_COMMIT_FORMATS = {
  CODE: 'code',
  EMOJI: 'emoji'
}

export default () => [
  {
    name: CONFIGURATION_PROMPT_NAMES.AUTO_ADD,
    message: 'Enable automatic "git add ."',
    type: 'confirm',
    default: configurationVault.getAutoAdd()
  },
  {
    name: CONFIGURATION_PROMPT_NAMES.EMOJI_FORMAT,
    message: 'Select how emojis should be used in commits',
    type: 'list',
    choices: [
      { name: ':smile:', value: EMOJI_COMMIT_FORMATS.CODE },
      { name: 'ðŸ˜„', value: EMOJI_COMMIT_FORMATS.EMOJI }
    ],
    default: configurationVault.getEmojiFormat()
  },
  {
    name: CONFIGURATION_PROMPT_NAMES.SIGNED_COMMIT,
    message: 'Enable signed commits',
    type: 'confirm',
    default: configurationVault.getSignedCommit()
  },
  {
    name: CONFIGURATION_PROMPT_NAMES.SCOPE_PROMPT,
    message: 'Enable scope prompt',
    type: 'confirm',
    default: configurationVault.getScopePrompt()
  }
]
