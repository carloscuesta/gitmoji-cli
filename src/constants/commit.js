// @flow
const COMMIT_MODES = {
  CLIENT: 'client',
  HOOK: 'hook'
}

export default COMMIT_MODES
