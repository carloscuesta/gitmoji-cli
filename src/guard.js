const chalk = require('chalk')

const constants = require('./constants')

const errors = {
  title: chalk.red('Enter a valid commit title'),
  message: chalk.red('Enter a valid commit message'),
  referenceGithub: chalk.red(
    'Enter the number of the reference without the #. Eg: 12'
  ),
  referenceJira: chalk.red('Enter the JIRA reference key, such as ABC-123')
}

const title = (title) => (!title || title.includes('`')) ? errors.title : true

const message = (message) => message.includes('`') ? errors.message : true

/**
* Checks that the given github reference is valid
* https://help.github.com/articles/autolinked-references-and-urls/
* @param {Number} ref - Represents a github reference
* @returns {Boolean}
*/
const githubGuard = (ref) => ref.match(/(^[1-9][0-9]*)+$/)
  ? true
  : errors.referenceGithub

/**
* Checks that the given jira reference is valid
* https://confluence.atlassian.com/jirasoftwarecloud/linking-issues-776997756.html
* @param {Number} ref - Represents a jira reference
* @returns {Boolean}
*/
const jiraGuard = (ref) => ref.match(/^([A-Z][A-Z0-9]{1,9}-[0-9]+)$/g)
  ? true
  : errors.referenceJira

const reference = (reference, mode) => {
  if (!reference) return true
  switch (mode) {
    case constants.GITHUB: {
      return githubGuard(reference)
    }

    case constants.JIRA: {
      return jiraGuard(reference)
    }

    default: {
      return githubGuard(reference)
    }
  }
}

module.exports = {
  title,
  message,
  reference
}
