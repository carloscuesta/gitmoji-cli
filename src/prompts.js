const constants = require('./constants')
const configVault = require('./config')
const guard = require('./guard')
const utils = require('./utils')

const config = [
  {
    name: constants.AUTO_ADD,
    message: 'Enable automatic "git add ."',
    type: 'confirm'
  },
  {
    name: constants.ISSUE_FORMAT,
    message: 'Choose Issue Format',
    type: 'list',
    choices: ['github', 'jira']
  },
  {
    name: constants.EMOJI_FORMAT,
    message: 'Select how emojis should be used in commits',
    type: 'list',
    choices: [
      { name: ':smile:', value: 'code' }, { name: 'ðŸ˜„', value: 'emoji' }
    ]
  },
  {
    name: constants.SIGNED_COMMIT,
    message: 'Enable signed commits',
    type: 'confirm'
  }
]

const gitmoji = (gitmojis) => {
  return [
    {
      name: 'gitmoji',
      message: 'Choose a gitmoji:',
      type: 'autocomplete',
      source: (answersSoFor, input) => {
        return Promise.resolve(
          gitmojis.filter((gitmoji) => {
            const emoji = gitmoji.name.concat(gitmoji.description).toLowerCase()
            return (!input || emoji.indexOf(input.toLowerCase()) !== -1)
          })
            .map((gitmoji) => ({
              name: `${gitmoji.emoji}  - ${gitmoji.description}`,
              value: gitmoji[configVault.getEmojiFormat() || constants.CODE]
            }))
        )
      }
    },
    {
      name: 'title',
      message: 'Enter the commit title',
      validate: guard.title,
      transformer: (input) => utils.inputCountTransformer(
        input,
        constants.TITLE_MAX_LENGTH_COUNT
      )
    },
    {
      name: 'message',
      message: 'Enter the commit message:',
      validate: guard.message
    },
    {
      name: 'reference',
      message: 'Issue / PR reference:',
      validate: (value) => guard.reference(value, configVault.getIssueFormat())
    }
  ]
}

module.exports = {
  config,
  gitmoji
}
