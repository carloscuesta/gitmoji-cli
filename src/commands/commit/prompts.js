// @flow
import inquirer from 'inquirer'

import configurationVault from '../../utils/configurationVault'
import guard from './guard'

const TITLE_MAX_LENGTH_COUNT: number = 48

inquirer.registerPrompt('autocomplete', require('inquirer-autocomplete-prompt'))

type Gitmoji = {
  code: string,
  description: string,
  emoji: string,
  name: string
}

export type Answers = {
  gitmoji: string,
  scope?: string,
  title: string,
  message: string
}

export default (gitmojis: Array<Gitmoji>): Array<Object> => [
  {
    name: 'gitmoji',
    message: 'Choose a gitmoji:',
    type: 'autocomplete',
    source: (answersSoFor: any, input: string) => {
      return Promise.resolve(
        gitmojis
          .filter((gitmoji) => {
            const emoji = gitmoji.name.concat(gitmoji.description).toLowerCase()
            return !input || emoji.indexOf(input.toLowerCase()) !== -1
          })
          .map((gitmoji) => ({
            name: `${gitmoji.emoji}  - ${gitmoji.description}`,
            value: gitmoji[configurationVault.getEmojiFormat()]
          }))
      )
    }
  },
  ...(configurationVault.getScopePrompt()
    ? [
        {
          name: 'scope',
          message: 'Enter the scope of current changes:',
          validate: guard.scope
        }
      ]
    : []),
  {
    name: 'title',
    message: 'Enter the commit title',
    validate: guard.title,
    transformer: (input: string) => {
      return `[${input.length}/${TITLE_MAX_LENGTH_COUNT}]: ${input}`
    }
  },
  {
    name: 'message',
    message: 'Enter the commit message:',
    validate: guard.message
  }
]
