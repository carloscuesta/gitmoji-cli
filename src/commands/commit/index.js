// @flow
import inquirer from 'inquirer'

import getEmojis from '../../utils/getEmojis'
import prompts from './prompts'
import withHook from './withHook'
import withClient from './withClient'

const commit = (mode: 'client' | 'hook') => {
  return getEmojis()
    .then((gitmojis) => prompts(gitmojis))
    .then((questions) => {
      inquirer.prompt(questions).then((answers) => {
        if (mode === 'hook') return withHook(answers)

        return withClient(answers)
      })
    })
}

export default commit
