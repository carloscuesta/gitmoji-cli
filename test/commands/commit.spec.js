import inquirer from 'inquirer'
import execa from 'execa'
import fs from 'fs'
const mockProcess = require('jest-mock-process')

import configurationVault from '../../src/utils/configurationVault'
import getEmojis from '../../src/utils/getEmojis'
import isHookCreated from '../../src/utils/isHookCreated'
import commit from '../../src/commands/commit'
import guard from '../../src/commands/commit/guard'
import prompts from '../../src/commands/commit/prompts'
import * as stubs from './stubs'

jest.mock('../../src/utils/getEmojis')
jest.mock('../../src/utils/isHookCreated')
jest.mock('../../src/utils/configurationVault')

describe('commit command', () => {
  describe('withClient', () => {
    describe('with no autoAdd and no signed commits and no scope', () => {
      beforeAll(() => {
        console.log = jest.fn()
        execa.mockReturnValue({ stdout: stubs.commitResult })
        inquirer.prompt.mockReturnValue(
          Promise.resolve(stubs.clientCommitAnswers)
        )
        getEmojis.mockResolvedValue(stubs.gitmojis)
        isHookCreated.mockResolvedValue(false)
        commit('client')
      })

      it('should call inquirer with prompts', () => {
        expect(inquirer.prompt.mock.calls).toMatchSnapshot()
      })

      it('should call execa with the commit command based on answers', () => {
        expect(execa).toHaveBeenCalledWith('git', [
          'commit',
          '-m',
          `${stubs.clientCommitAnswers.gitmoji} ${stubs.clientCommitAnswers.title}`,
          '-m',
          stubs.clientCommitAnswers.message
        ])
      })

      it('should print the result to the console', () => {
        expect(console.log).toHaveBeenCalledWith(stubs.commitResult)
      })
    })

    describe('with autoAdd, signed commits and scope', () => {
      beforeAll(() => {
        console.log = jest.fn()
        execa.mockReturnValue({ stdout: stubs.commitResult })
        inquirer.prompt.mockReturnValue(
          Promise.resolve(stubs.clientCommitAnswersWithScope)
        )
        getEmojis.mockResolvedValue(stubs.gitmojis)
        isHookCreated.mockResolvedValue(false)
        configurationVault.getAutoAdd.mockReturnValue(true)
        configurationVault.getSignedCommit.mockReturnValue(true)
        commit('client')
      })

      it('should call inquirer with prompts', () => {
        expect(inquirer.prompt.mock.calls).toMatchSnapshot()
      })

      it('should call execa with the add command', () => {
        expect(execa).toHaveBeenCalledWith('git', ['add', '.'])
      })

      it('should call execa with the commit command based on answers', () => {
        expect(execa).toHaveBeenLastCalledWith('git', [
          'commit',
          '-S',
          '-m',
          `${stubs.clientCommitAnswersWithScope.gitmoji} (${stubs.clientCommitAnswersWithScope.scope}): ${stubs.clientCommitAnswersWithScope.title}`,
          '-m',
          stubs.clientCommitAnswersWithScope.message
        ])
      })

      it('should print the result to the console', () => {
        expect(console.log).toHaveBeenCalledWith(stubs.commitResult)
      })
    })

    describe('with the commit hook created', () => {
      beforeAll(() => {
        console.log = jest.fn()
        inquirer.prompt.mockReturnValue(
          Promise.resolve(stubs.clientCommitAnswers)
        )
        getEmojis.mockResolvedValue(stubs.gitmojis)
        isHookCreated.mockResolvedValue(true)
        commit('client')
      })

      it('should call inquirer with prompts', () => {
        expect(inquirer.prompt.mock.calls).toMatchSnapshot()
      })

      it('should stop the commit because the hook is created and log the explanation to the user', () => {
        expect(console.log).toHaveBeenCalledWith(expect.any(String))
        expect(execa).not.toHaveBeenCalledWith()
      })
    })
  })

  describe('withHook', () => {
    describe('without scope', () => {
      beforeAll(() => {
        console.log = jest.fn()
        inquirer.prompt.mockReturnValue(
          Promise.resolve(stubs.clientCommitAnswers)
        )
        getEmojis.mockResolvedValue(stubs.gitmojis)
        mockProcess.mockProcessExit()
        process.argv[3] = stubs.argv
        commit('hook')
      })

      it('should commit using the hook', () => {
        expect(fs.writeFileSync).toHaveBeenCalledWith(
          stubs.argv,
          `${stubs.clientCommitAnswers.gitmoji} ${stubs.clientCommitAnswers.title}\n\n${stubs.clientCommitAnswers.message}`
        )
      })

      it('should call process.exit', () => {
        expect(process.exit).toHaveBeenCalledWith(0)
      })
    })

    describe('with scope', () => {
      beforeAll(() => {
        console.log = jest.fn()
        inquirer.prompt.mockReturnValue(
          Promise.resolve(stubs.clientCommitAnswersWithScope)
        )
        getEmojis.mockResolvedValue(stubs.gitmojis)
        mockProcess.mockProcessExit()
        process.argv[3] = stubs.argv
        commit('hook')
      })

      it('should commit using the hook', () => {
        expect(fs.writeFileSync).toHaveBeenCalledWith(
          stubs.argv,
          `${stubs.clientCommitAnswersWithScope.gitmoji} (${stubs.clientCommitAnswersWithScope.scope}): ${stubs.clientCommitAnswersWithScope.title}\n\n${stubs.clientCommitAnswersWithScope.message}`
        )
      })

      it('should call process.exit', () => {
        expect(process.exit).toHaveBeenCalledWith(0)
      })
    })
  })

  describe('guard', () => {
    it('should match guard', () => {
      expect(guard).toMatchSnapshot()
    })

    describe('title', () => {
      it('should return true when valid', () => {
        expect(guard.title(stubs.commitTitle)).toBe(true)
      })

      it('should return error message when empty', () => {
        expect(guard.title('')).toEqual(expect.any(String))
      })

      it('should return error message with invalid characters', () => {
        expect(guard.title(stubs.commitTitleInvalid)).toEqual(
          expect.any(String)
        )
      })
    })

    describe('message', () => {
      it('should return true when valid', () => {
        expect(guard.title(stubs.commitTitle)).toBe(true)
      })

      it('should return error message when empty', () => {
        expect(guard.title('')).toEqual(expect.any(String))
      })

      it('should return error message with invalid characters', () => {
        expect(guard.title(stubs.commitTitleInvalid)).toEqual(
          expect.any(String)
        )
      })
    })

    describe('scope', () => {
      it('should return true when valid', () => {
        expect(guard.title(stubs.commitTitle)).toBe(true)
      })

      it('should return error message when empty', () => {
        expect(guard.title('')).toEqual(expect.any(String))
      })

      it('should return error message with invalid characters', () => {
        expect(guard.title(stubs.commitTitleInvalid)).toEqual(
          expect.any(String)
        )
      })
    })
  })

  describe('prompts', () => {
    it('should register the autoComplete inquirer prompt', () => {
      expect(inquirer.registerPrompt).toHaveBeenCalledWith(
        'autocomplete',
        expect.any(Function)
      )
    })

    describe('without scope prompt', () => {
      it('should match the array of questions', () => {
        expect(prompts(stubs.gitmojis)).toMatchSnapshot()
      })
    })

    describe('with scope prompt', () => {
      beforeAll(() => {
        configurationVault.getScopePrompt.mockReturnValue(true)
      })

      it('should match the array of questions', () => {
        expect(prompts(stubs.gitmojis)).toMatchSnapshot()
      })
    })
  })
})
