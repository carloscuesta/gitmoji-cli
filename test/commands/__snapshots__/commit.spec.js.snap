// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commit command guard should match guard 1`] = `
Object {
  "message": [Function],
  "scope": [Function],
  "title": [Function],
}
`;

exports[`commit command prompts with scope prompt should match the array of questions 1`] = `
Array [
  Object {
    "message": "Choose a gitmoji:",
    "name": "gitmoji",
    "source": [Function],
    "type": "autocomplete",
  },
  Object {
    "message": "Enter the scope of current changes:",
    "name": "scope",
    "validate": [Function],
  },
  Object {
    "message": "Enter the commit title",
    "name": "title",
    "transformer": [Function],
    "validate": [Function],
  },
  Object {
    "message": "Enter the commit message:",
    "name": "message",
    "validate": [Function],
  },
]
`;

exports[`commit command prompts without scope prompt should match the array of questions 1`] = `
Array [
  Object {
    "message": "Choose a gitmoji:",
    "name": "gitmoji",
    "source": [Function],
    "type": "autocomplete",
  },
  Object {
    "message": "Enter the commit title",
    "name": "title",
    "transformer": [Function],
    "validate": [Function],
  },
  Object {
    "message": "Enter the commit message:",
    "name": "message",
    "validate": [Function],
  },
]
`;

exports[`commit command withClient with autoAdd, signed commits and scope should call inquirer with prompts 1`] = `
Array [
  Array [
    Array [
      Object {
        "message": "Choose a gitmoji:",
        "name": "gitmoji",
        "source": [Function],
        "type": "autocomplete",
      },
      Object {
        "message": "Enter the commit title",
        "name": "title",
        "transformer": [Function],
        "validate": [Function],
      },
      Object {
        "message": "Enter the commit message:",
        "name": "message",
        "validate": [Function],
      },
    ],
  ],
  Array [
    Array [
      Object {
        "message": "Choose a gitmoji:",
        "name": "gitmoji",
        "source": [Function],
        "type": "autocomplete",
      },
      Object {
        "message": "Enter the commit title",
        "name": "title",
        "transformer": [Function],
        "validate": [Function],
      },
      Object {
        "message": "Enter the commit message:",
        "name": "message",
        "validate": [Function],
      },
    ],
  ],
]
`;

exports[`commit command withClient with no autoAdd and no signed commits and no scope should call inquirer with prompts 1`] = `
Array [
  Array [
    Array [
      Object {
        "message": "Choose a gitmoji:",
        "name": "gitmoji",
        "source": [Function],
        "type": "autocomplete",
      },
      Object {
        "message": "Enter the commit title",
        "name": "title",
        "transformer": [Function],
        "validate": [Function],
      },
      Object {
        "message": "Enter the commit message:",
        "name": "message",
        "validate": [Function],
      },
    ],
  ],
]
`;

exports[`commit command withClient with the commit hook created should call inquirer with prompts 1`] = `
Array [
  Array [
    Array [
      Object {
        "message": "Choose a gitmoji:",
        "name": "gitmoji",
        "source": [Function],
        "type": "autocomplete",
      },
      Object {
        "message": "Enter the commit title",
        "name": "title",
        "transformer": [Function],
        "validate": [Function],
      },
      Object {
        "message": "Enter the commit message:",
        "name": "message",
        "validate": [Function],
      },
    ],
  ],
  Array [
    Array [
      Object {
        "message": "Choose a gitmoji:",
        "name": "gitmoji",
        "source": [Function],
        "type": "autocomplete",
      },
      Object {
        "message": "Enter the commit title",
        "name": "title",
        "transformer": [Function],
        "validate": [Function],
      },
      Object {
        "message": "Enter the commit message:",
        "name": "message",
        "validate": [Function],
      },
    ],
  ],
  Array [
    Array [
      Object {
        "message": "Choose a gitmoji:",
        "name": "gitmoji",
        "source": [Function],
        "type": "autocomplete",
      },
      Object {
        "message": "Enter the commit title",
        "name": "title",
        "transformer": [Function],
        "validate": [Function],
      },
      Object {
        "message": "Enter the commit message:",
        "name": "message",
        "validate": [Function],
      },
    ],
  ],
]
`;
